crates/wasmtime/src/linker.rs

```diff
@@ -270,6 +271,117 @@ impl Linker {
+    /// let module = Module::new(&store, wat)?;
+    /// linker.module("instance1", &module)?;
+    ///
+    /// let wat = r#"
+    ///     (module
+    ///         (import "instance1" "run" (func $instance1_run))
+    ///         (func (export "run")
+    ///             call $instance1_run
+    ///         )
+    ///     )
+    /// "#;
+    /// let module = Module::new(&store, wat)?;
+    /// let instance = linker.instantiate(&module)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn module(&mut self, module_name: &str, module: &Module) -> Result<&mut Self> {
+        match ModuleKind::categorize(module)? {
+            ModuleKind::Command => self.command(module_name, module),
+            ModuleKind::Reactor => {
+                let instance = self.instantiate(&module)?.start()?;
+
+                if let Some(export) = instance.get_export("_initialize") {
+                    if let Extern::Func(func) = export {
+                        func.get0::<()>()
+                            .and_then(|f| f().map_err(Into::into))
+                            .context("calling the Reactor initialization function")?;
+                    }
+                }
+
+                self.instance(module_name, &instance)
+            }
+        }
+    }
+
+    fn command(&mut self, module_name: &str, module: &Module) -> Result<&mut Self> {
+        for export in module.exports() {
+            if let Some(func_ty) = export.ty().func() {
+                let imports = self.compute_imports(module)?;
+                let module = module.clone();
+                let export_name = export.name().to_owned();
+                let func = Func::new(&self.store, func_ty.clone(), move |_, params, results| {
+                    let instance = Instance::new(&module, &imports)
+                        .map_err(|error| match error.downcast::<Trap>() {
+                            Ok(trap) => trap,
+                            Err(error) => Trap::new(error.to_string()),
+                        })?
+                        .start()?;
+                    let command_results = instance
+                        .get_export(&export_name)
+                        .unwrap()
+                        .into_func()
+                        .unwrap()
+                        .call(params)
+                        .map_err(|error| error.downcast::<Trap>().unwrap())?;
+                    for (result, command_result) in
+                        results.iter_mut().zip(command_results.into_vec())
+                    {
+                        *result = command_result;
+                    }
+                    Ok(())
+                });
+                self.insert(module_name, export.name(), Extern::Func(func))?;
+            } else if export.name() == "memory" && export.ty().memory().is_some() {
+                // Allow an exported "memory" memory for now.
+            } else if export.name() == "__indirect_function_table" && export.ty().table().is_some()
```

Comment:
These and below are definitely a bummer, but I guess there's not much we can do about it?

How come though we validate that only functions and a few whitelisted things are exported?

Reply:
Rust is one of the toolchains using `--export-dynamic` and I'd ideally like to remove that, but I need to figure out exactly how Rust use cases are depending on that before we can remove it. I've heard of other toolchains using it as well, as it sort of gives people what they want -- "just export my stuff". Except that `--export-dynamic` was implemented in wasm-ld to handle the Emscripten-specific dynamic linking scheme used in Emscripten, so it also does some additional things that Emscripten needs like exporting `__heap_base` and `__data_end`, even when we aren't using Emscripten. So we have a multi-project backwards compatibility knot to untie before we can fix this.

In general, it doesn't make sense for Commands to export globals, tables, or memories, because those are all about state, and Commands don't have any state, except when one of their exported functions is being called.

Reply:
Ah yeah our usage of LLD is *very* dated at this point in rustc. At least for rustc it probably won't be too too hard to fix this I think.